torchAPI.lua

```diff
@@ -0,0 +1,215 @@
+  
+  <tag>    <rtype>   <library>_<pattern>
+  TH_API     void  THNN_Abs_updateOutput
+
+Equivalently, if we initialize an API with
+pattern := '%(([%a%d_]+)%)':
+  
+  <tag>    <rtype>    <library>_<pattern>
+  TH_API    void    THNN_(Abs_updateOutput)
+
+
+Authored: 2016-01-05 (jwilson)
+Modified: 2016-02-04
+--]]
+
+---------------- External Dependencies
+local ffi = require 'ffi'
+
+------------------------------------------------
+--                                           API
+------------------------------------------------
+local API = torch.class('torchAPI')
+
+function API:__init(config, header)
+  local L = config or {}
+  assert(type(L.library) == 'string')
+  L.pattern = L.pattern or '([%a%d_]+)'
+  L.rtype   = L.rtype   or 'void'
+  L.tag     = L.tag     or ''
+  L.prefix  = L.prefix  or string.format('%s %s %s_', L.tag, L.rtype, L.library):gsub("^%s+", "")
+  self.config = L 
+
+  if header then
+    self:c_init(header)
+  end
+end
+
+function API:getState(ttype)
+  if ttype == 'Cuda' then
+    if not self.cuda_ptr then 
+      self.cuda_ptr = ffi.typeof('THCState*')
+    end
+    return self.cuda_ptr(cutorch.getState())
+  else
+    return ffi.NULL or nil
+  end
+end
+
+function API:extract_handles(forward, pattern, handles)
+  local handles = handles or {}
+  local pattern = pattern or self.config.prefix ..self.config.pattern
+  for handle in string.gmatch(forward, pattern) do
+    handles[#handles+1] = handle
+  end
+  return handles
+end
+
+function API:c_init(header)
+
+  ---- Define C Preamble
+  if header['preamble'] then
+    ffi.cdef(header['preamble'])
+  end
+
+  ---- Check for forward declaration
+  forward = header['forward']
+  if not forward then return end -- terminate if no forward
+
+  ---- Swap in template Prepare forward for declaration / expansion of macros 
+  local templates = header['template']
+  if templates then
+    if type(templates) == 'string' then
+      templates = {templates}
+    end
+
+    for idx = 1, #templates do
+      template = templates[idx]
+
+      ---- Check for special patterns
+      if template:match('<tag>') then
+        template = template:gsub('<tag>', self.config.tag)
+      end
+
+      if template:match('<rtype>') then
+        template = template:gsub('<rtype>', self.config.rtype)
+      end
+
+      if template:match('<library>') then
+        template = template:gsub('<library>', self.config.library)
+      end
+
+      if template:match('$([0-9]+)') then
+        template = template:gsub('$([0-9]+)', '%%%1')
+      end
+
+      forward = forward:gsub(self.config.prefix..self.config.pattern, template)
+    end
+  end
+
+  ---- Expand macros
+  local macros = header['macro'] or header['macros']
+  if macros then
+    local macros = self:_interp_macros(macros)
+    local flag   = false
+
+    -- Apply macros with same interpretation everywhere
+    for old, new in pairs(macros) do
+      if type(new) == 'string' then
+        forward = forward:gsub(old, new)
```

Comment:
It's not exactly how C macros work (this will also replace subwords and string contents), but I'm not sure if it's a problem when parsing only the header files.